javalib/src/main/scala/java/util/Collections.scala

```diff
@@ -0,0 +1,1216 @@
+    unmodifiableMap[K, V](new AbstractMap[K, V] with Serializable {
+      def entrySet(): Set[Map.Entry[K, V]] =
+        singleton(new AbstractMap.SimpleImmutableEntry(key, value))
+    })
+  }
+
+  def nCopies[T](n: Int, o: T): List[T] = {
+    if (n < 0)
+      throw new IllegalArgumentException
+
+    unmodifiableList(new AbstractList[T] with Serializable with RandomAccess {
+      def size(): Int = n
+
+      def get(index: Int): T = {
+        if (index < 0 || index >= n)
+          throw new IndexOutOfBoundsException
+        o
+      }
+    })
+  }
+
+  def reverseOrder[T](): Comparator[T] = {
+    val cmp = new Comparator[T] {
+      def compare(o1: T, o2: T): Int = o1.asInstanceOf[Comparable[T]].compareTo(o2)
+    }
+    reverseOrder(cmp)
+  }
+
+  def reverseOrder[T](cmp: Comparator[T]): Comparator[T] = {
+    class ReverseComparator extends Comparator[T] {
+      override def compare(o1: T, o2: T): Int = cmp.compare(o2, o1)
+    }
+    cmp match {
+      case _: Serializable => new ReverseComparator with Serializable
+      case _               => new ReverseComparator
+    }
+  }
+
+  def enumeration[T](c: Collection[T]): Enumeration[T] = {
+    val it = c.iterator
+    new Enumeration[T] {
+      override def hasMoreElements: Boolean =
+        it.hasNext
+
+      override def nextElement(): T =
+        it.next()
+    }
+  }
+
+  def list[T](e: Enumeration[T]): ArrayList[T] = {
+    val arrayList = new ArrayList[T]
+    e.foreach(arrayList += _)
+    arrayList
+  }
+
+  def frequency(c: Collection[_], o: AnyRef): Int =
+    c.count(_ === o)
+
+  def disjoint(c1: Collection[_], c2: Collection[_]): Boolean = {
+    if (c1.size < c2.size)
+      !c1.exists(elem => c2.contains(elem))
+    else
+      !c2.exists(elem => c1.contains(elem))
+  }
+
+  def addAll[T](c: Collection[_ >: T], elements: Array[AnyRef]): Boolean =
+    c.addAll(asJavaCollection(elements.asInstanceOf[Array[T]]))
+
+  def newSetFromMap[E](map: Map[E, java.lang.Boolean]): Set[E] = {
+    if (map.nonEmpty)
+      throw new IllegalArgumentException
+
+    new WrappedSet[E, Set[E]] {
+      override protected val inner: Set[E] =
+        map.keySet
+
+      override def add(e: E): Boolean =
+        map.put(e, true) == null
+
+      override def addAll(c: Collection[_ <: E]): Boolean =
+        c.foldLeft(false)((prev, elem) => map.put(elem, true) == null || prev)
+    }
+  }
+
+  @inline
+  private def modulo(a: Int, b: Int): Int = ((a % b) + b) % b
+
+  @inline
+  private def naturalComparator[T <: jl.Comparable[T]]: Comparator[T] = {
+    new Comparator[T] {
+      final def compare(o1: T, o2: T): Int = o1.compareTo(o2)
+    }
+  }
+
+  @inline
+  private implicit def comparatorToOrdering[E](cmp: Comparator[E]): Ordering[E] = {
+    new Ordering[E] {
+      final def compare(x: E, y: E): Int = cmp.compare(x, y)
+    }
+  }
+
+  private object WrappedCollections {
+
+    private[Collections] trait WrappedCollection[E, Coll <: Collection[E]] extends Collection[E] {
```

Comment:
Could all these things be classes instead of traits? Even better, with `inner` as a constructor parameter?
This would reduce dramatically the amount of generated code.

Btw, isn't the nesting in `WrappedCollections` superfluous? Especially since a) the traits are named WrappedXYZ and b) you import the contents of the three "namespaces" at the top of the file anyway.